chia/wallet/coin_selection.py

```diff
@@ -73,23 +74,32 @@ async def select_coins(
         if coin.amount < amount:
             smaller_coin_sum += coin.amount
             smaller_coins.append(coin)
-    if smaller_coin_sum == amount:
+    if smaller_coin_sum == amount and len(smaller_coins) < max_num_coins:
         log.debug(f"Selected all smaller coins because they equate to an exact match of the target.: {smaller_coins}")
         return set(smaller_coins)
     elif smaller_coin_sum < amount:
-        if len(smaller_coins) > 0:  # in case we only have bigger coins.
-            greater_coins = valid_spendable_coins[: -len(smaller_coins)]
-        else:
-            greater_coins = valid_spendable_coins
-        smallest_coin = greater_coins[len(greater_coins) - 1]  # select the coin with the least value.
-        assert smallest_coin is not None
+        smallest_coin = select_smallest_coin_over_target(len(smaller_coins), valid_spendable_coins)
         log.debug(f"Selected closest greater coin: {smallest_coin.name()}")
         return {smallest_coin}
+    elif smaller_coin_sum > amount:
+        coin_set = knapsack_coin_algorithm(smaller_coins, amount, max_coin_amount)
+        log.debug(f"Selected coins from knapsack algorithm: {coin_set}")
+        if coin_set is None:
+            raise ValueError("Knapsack algorithm failed to find a solution.")
+        if len(coin_set) > max_num_coins:
+            coin = select_smallest_coin_over_target(len(smaller_coins), valid_spendable_coins)
+            if coin is None or coin.amount < amount:
+                raise ValueError(
+                    "Can't make this transaction because the transaction would use over 500 "

```

Comment:
Suggest more concise string `raise ValueError(f"Transaction of {amount} would use more than 500 coins. Try sending a smaller amount")